packages/react/toggle-button/src/ToggleButton.tsx

```diff
@@ -57,12 +247,138 @@ const ToggleButton = forwardRefWithAs<typeof DEFAULT_TAG, ToggleButtonOwnProps>(
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    React.useEffect(() => {
+      const propName = isControlled ? '`toggled`' : '`defaultToggled`';
+      const toggled = isControlled ? toggledProp : defaultToggled;
+      warning(
+        !(groupContext && toggled !== undefined),
+        `A ${BUTTON_NAME} with an explicit ${propName} prop was used inside of a ${groupName}. When ${BUTTON_NAME} is used inside of a ${groupName}, the ${groupName} component is responsible for managing the toggled state of its nested ${BUTTON_NAME} components. Either remove the ${propName} from ${BUTTON_NAME} or remove the ${groupName} component if the ${BUTTON_NAME} is unrelated to other surrounding buttons.`
+      );
+    }, [groupContext, toggledProp, defaultToggled, groupName, isControlled]);
+  }
+}
+
+function stringArraysAreEqual(arr1: string[] | undefined, arr2: string[] | undefined) {
+  if (arr1 === undefined && arr2 === undefined) return true;
+  if (arr1 === undefined) return false;
+  if (arr2 === undefined) return false;
+  return JSON.stringify(arr1.sort()) === JSON.stringify(arr2.sort());
+}
+
+// NOTE: I changed this implementation a tiny bit to improve typing somewhat. I think it's a little
+// nicer not to include `undefined` in the typing here and always provide a defaultProp value
+// whether or not its passed by the consumer. If undefined is an acceptable type it should be added
+// explicitly IMO. Also added an option to override the equality check so we can handle arrays,
+// objects, etc. If we're ok with these changes I can update it in the utils package as a separate
+// PR.
+function useControlledState<T>({
+  prop,
+  defaultProp,
+  onChange,
+  isEqual = defaultIsEqual,
+}: {
+  prop?: T;
+  defaultProp: T;
+  onChange?: (value: T) => void;
+  isEqual?: (prevState: any, nextState: any) => boolean;
+}): [T, React.Dispatch<React.SetStateAction<T>>] {
+  const stable_onChange = useCallbackRef(onChange);
+  const stable_isEqual = useCallbackRef(isEqual);
+  const [state, setState] = React.useState(prop || defaultProp);
+  const stateRef = React.useRef(state);
+  const isControlled = useIsControlled(prop);
+
+  const setValue: React.Dispatch<React.SetStateAction<T>> = React.useCallback(
+    (value) => {
+      function handleChange(value: T) {
+        if (!stable_isEqual(stateRef.current, value)) {
+          stable_onChange(value);
+        }
+        if (!isControlled) {
+          stateRef.current = value;
+        }
+      }
+
+      if (isFunction(value)) {
+        setState((oldValue) => {
+          const newValue = value((isControlled ? stateRef.current : oldValue) as any);
+          handleChange(newValue);
+          return isControlled ? oldValue : newValue;
+        });
+      } else {
+        if (!isControlled) {
+          setState(value);
+        }
+        handleChange(value);
+      }
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [stable_isEqual, stable_onChange]
+  );
```

Comment:
`isEqual` is there primarily to compare array values since the references change and, without it, it was firing when it shouldn't when I'd select a single value that couldn't be toggled off when `selectionIsRequired` was set to true. I'll revisit the implementation against what we had, but the only reason I pulled it here was to debug issues I was having with it that case without affecting other packages.

Reply:
> `isEqual` is there primarily to compare array values since the references change and, without it, it was firing when it shouldn't

Yep, that's what the new `useControlledState` hook in `main` solves. You can use the function callback for `setValue` to compare references before returning the new value. If a change hasn't occurred, you return the `prevValue` and it doesn't fire the change callback.

Reply:
I updated this to reflect the current implementation of `useControlledState` with only the changes needed for types. Figured we could carry on that conversation separately so I can mark this one resolved.